channeldb/migrations.go

```diff
@@ -934,3 +934,77 @@ func migrateOutgoingPayments(tx *bbolt.Tx) error {
 	log.Infof("Migration of outgoing payment bucket structure completed!")
 	return nil
 }
+
+// migrateRouteSerialization migrates the way we serialize routes across the
+// entire database. At the time of writing of this migration, this includes our
+// payment attempts, as well as the payment result sin mission control.
+func migrateRouteSerialization(tx *bbolt.Tx) error {
+	// First, we'll do all the payment attempts.
+	rootPaymentBucket := tx.Bucket(paymentsRootBucket)
+	if rootPaymentBucket == nil {
+		return nil
+	}
+
+	// As we can't mutate a bucket while we're iterating over it with
+	// ForEach, we'll need to collect all the known payment hashes in
+	// memory first.
+	var payHashes [][]byte
+	err := rootPaymentBucket.ForEach(func(k, v []byte) error {
+		if v != nil {
+			return nil
+		}
+
+		payHash := make([]byte, len(k))
+		copy(payHash, k)
+
+		payHashes = append(payHashes, payHash)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Now that we have all the payment hashes, we can carry out the
+	// migration itself.
+	for _, payHash := range payHashes {
+		payHashBucket := rootPaymentBucket.Bucket(payHash)
+		payAttemptBytes := payHashBucket.Get(paymentAttemptInfoKey)
+		if payAttemptBytes == nil {
+			continue
+		}
+
+		// For our migration, we'll first read out the existing payment
+		// attempt using the legacy serialization of the attempt.
+		payAttemptReader := bytes.NewReader(payAttemptBytes)
+		payAttempt, err := deserializePaymentAttemptInfoLegacy(
+			payAttemptReader,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Now that we have the old attempts, we'll explicitly mark
+		// this as needing a legacy payload, since after this
+		// migration, the modern payload will be the default if
+		// signalled.
+		for _, hop := range payAttempt.Route.Hops {
+			hop.LegacyPayload = true
+		}
+
+		// Finally, we'll write out the payment attempt using the new
+		// encoding.
+		var b bytes.Buffer
+		err = serializePaymentAttemptInfo(&b, payAttempt)
+		if err != nil {
+			return err
+		}
+
+		err = payHashBucket.Put(paymentAttemptInfoKey, b.Bytes())
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Infof("Migration of route/hop serialization complete!")
```

Comment:
Need to clear mission control here as a minimum form of migration.

Reply:
Was planning on piggy backing off that other mission control migration (final cltv), but looks like that's been paused.

Reply:
I think we should _move_ the mission control store code into `channeldb` though, as then we can do a proper migration in the future, as they way it is now, we can't access any of the private methods.

Reply:
Hmm yes I should have placed it there indeed. Yes, the other mission control migration is delayed. It is based of the invoice htlc tracking pr #3390, so that will take longer to land.